### PR TITLE
topology: pre-process-object: Expand definitions within strings

### PR DESCRIPTION
**Pull request by jsarha**
The second commit description tries to explain what we are doing here. I was in a little bit of hurry when finishing this. I hope everything is in order. At least this should be good enough to start a discussion.

I'll be away from tomorrow afternoon until 2nd of Feb, but @ranj063 should be able to help with this.

Best regards,
Jyri

### CODE REVIEW

**Comment by jsarha** [comment]
@perexg , how does this look?

File: topology/pre-process-object.c [hunk 1]
```diff
@@ -1610,6 +1610,173 @@ pre_process_object_variables_expand_fcn(snd_config_t **dst, const char *str, voi
...
+{
+	size_t prefix_len, varname_len;
+
+	*prefix = NULL;
+	*varname = NULL;
+
+	prefix_len = strcspn(*stringp, "$");
+	*prefix = *stringp;
+	(*stringp) += prefix_len;
+	varname_len = strcspn(*stringp, " .-");
```
[the remark this answers is not in the document]
Comment (perexg) [reply topology/pre-process-object.c:1642]: I would use `isalnum()` function + `_` (underscore) character to check for the end-of-variable-name to be consistent with https://github.com/alsa-project/alsa-lib/blob/master/src/confeval.c .

Also, `$[VARNAME]` notation may be handled for variables containing other characters.

File: topology/pre-process-object.c [hunk 2]
```diff
@@ -1610,6 +1610,173 @@ pre_process_object_variables_expand_fcn(snd_config_t **dst, const char *str, voi
...
+				goto out;
+
+			if (*dst == NULL) {
+				*dst = tmp_config;
+			} else {
+				snd_config_type_t type = snd_config_get_type(tmp_config);
+
+				snd_config_get_string(*dst, &current_str);
+				/* concat the value based on type to the current string */
+				switch (type) {
```
Comment (perexg) [inline topology/pre-process-object.c:1731]: Use `snd_config_get_ascii()` function here...

File: topology/pre-process-object.c [hunk 3]
```diff
@@ -1610,6 +1610,173 @@ pre_process_object_variables_expand_fcn(snd_config_t **dst, const char *str, voi
...
+	char *varname, *prefix, *freep = str;
+	int ret;
+
+	if (!str)
+		return -ENOMEM;
+
+	*dst = NULL;
+
+	/* evaluate the whole string the string is an arthmetic expression */
+	if ((s[0] == '$' && s[1] == '[')) {
```
Comment (perexg) [inline topology/pre-process-object.c:1677]: Double `((` and `))` ???

**Review by perexg — CHANGES_REQUESTED** [review]

**Comment by jsarha** [comment]
Thanks @perexg for your comments. This new version should address all the comments, yours and mine:
- Handle also $[$VAR_NAME + 1] notation within strings
- Handle anything but alpha numeric and '_' as a delimiter for variable name
- Use snd_config_get_ascii()
- Fix error print in case tplg_evaluate_config_string() fails

**Comment by perexg** [comment]
Thanks. Merged.

**Pull request closed without merging**